File: market/canonical/canonical.go
// Package canonical provides deterministic ordering and checksum primitives for
// replayable market-calculation inputs.
package canonical

import (
	"crypto/sha256"
	"encoding/binary"
	"sort"
)

// SortedStrings returns a sorted copy of values. It never mutates caller input.
func SortedStrings(values []string) []string {
	result := append([]string(nil), values...)
	sort.Strings(result)
	return result
}

// Checksum hashes an unambiguous length-prefixed sequence of canonical parts.
// Callers must supply values in their documented canonical order.
func Checksum(parts ...string) [sha256.Size]byte {
	hash := sha256.New()
	var length [8]byte
	for _, part := range parts {
		binary.BigEndian.PutUint64(length[:], uint64(len(part)))
		_, _ = hash.Write(length[:])
		_, _ = hash.Write([]byte(part))
	}
	var result [sha256.Size]byte
	copy(result[:], hash.Sum(nil))
	return result
}

File: src/market/canonical.ts
/** Deterministic ordering and checksum primitives (ported from market/canonical). */

import { bytesToHex } from "../utils/crypto.ts";

/** Returns a sorted copy of values without mutating caller input. */
export function sortedStrings(values: string[]): string[] {
  return [...values].sort();
}

/**
 * Hashes an unambiguous length-prefixed sequence of canonical parts and
 * returns the hex-encoded SHA-256 digest.
 */
export async function canonicalChecksum(parts: string[]): Promise<string> {
  const encoder = new TextEncoder();
  const chunks: Uint8Array[] = [];
  const lengthBuffer = new Uint8Array(8);
  const dataView = new DataView(lengthBuffer.buffer);
  for (const part of parts) {
    dataView.setBigUint64(0, BigInt(part.length), false);
    chunks.push(lengthBuffer);
    chunks.push(encoder.encode(part));
  }
  const totalLength = chunks.reduce((sum, chunk) => sum + chunk.length, 0);
  const combined = new Uint8Array(totalLength);
  let offset = 0;
  for (const chunk of chunks) {
    combined.set(chunk, offset);
    offset += chunk.length;
  }
  const digest = await crypto.subtle.digest("SHA-256", combined);
  return bytesToHex(new Uint8Array(digest));
}
